protocols/gossipsub/src/types.rs

```diff
@@ -231,125 +244,151 @@ pub struct Rpc {
     pub control_msgs: Vec<ControlAction>,
 }
 
-impl Rpc {
+impl fmt::Debug for Rpc {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut b = f.debug_struct("GossipsubRpc");
+        if !self.messages.is_empty() {
+            b.field("messages", &self.messages);
+        }
+        if !self.subscriptions.is_empty() {
+            b.field("subscriptions", &self.subscriptions);
+        }
+        if !self.control_msgs.is_empty() {
+            b.field("control_msgs", &self.control_msgs);
+        }
+        b.finish()
+    }
+}
+
+/// A Gossipsub RPC message sent.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum RpcOut {
+    /// Publish a Gossipsub message on network.
+    Publish(RawMessage),
+    /// Forward a Gossipsub message to the network.
+    Forward(RawMessage),
+    /// Subscribe a topic.
+    Subscribe(TopicHash),
+    /// Unsubscribe a topic.
+    Unsubscribe(TopicHash),
+    /// List of Gossipsub control messages.
+    Control(Vec<ControlAction>),
+}
+
+impl RpcOut {
     /// Converts the GossipsubRPC into its protobuf format.
     // A convenience function to avoid explicitly specifying types.
     pub fn into_protobuf(self) -> proto::RPC {
```

Comment:
semver-checks complains that this function is now removed from the public `Rpc` type. You'll have to add it back to make this backwards compatible :)

Reply:
I don't think we need to with https://github.com/libp2p/rust-libp2p/pull/4833 on the way :)